### PR TITLE
AP_SteerController: Steer limit on speed

### PR DESCRIPTION
**Pull request by EShamaev**
AP_SteerController: added derating of steer angle depending on speed

When using steering controller on my plane (big - 30 kg, 3.6m wingspan) at high speed sometimes controller oversteers that leads to rolling on the wing.
Derating of steering was added.
Below certain speed - no derating, then as speed raises - maximum throw is lowered until defined minimum throw is reached.

### CODE REVIEW

File: libraries/APM_Control/AP_SteerController.cpp [hunk 1]
```diff
@@ -156,11 +184,21 @@ int32_t AP_SteerController::get_steering_out_rate(float desired_rate)
...
-	_last_out = _pid_info.D + _pid_info.FF + _pid_info.P + _pid_info.I;
+    // Calculate the demanded control surface deflection
+    _last_out = _pid_info.D + _pid_info.FF + _pid_info.P + _pid_info.I;
 	
-	// Convert to centi-degrees and constrain
-	return constrain_float(_last_out * 100, -4500, 4500);
+    float derate_constraint = 4500;
+
+    // Calculate required constrain based on speed
+    if(speed > _deratespeed) {
```
Comment (muramura) [inline libraries/APM_Control/AP_SteerController.cpp]: this if and other if statement , in white space.

Comment (EShamaev) [inline libraries/APM_Control/AP_SteerController.cpp]: Done.

File: libraries/APM_Control/AP_SteerController.cpp [hunk 2]
```diff
@@ -83,6 +83,34 @@ const AP_Param::GroupInfo AP_SteerController::var_info[] = {
...
+    AP_GROUPINFO("DRTFCT", 9, AP_SteerController, _deratefactor,        10),
+
+    // @Param: DRTMIN
+    // @DisplayName: Minimum angle of wheel
+    // @Description: The angle that limits smalles angle of steering wheel at maximum speed. Even if it should derate below, it will stop derating at this angle.
+    // @Range: 0.0 4500.0
+    // @Increment: 0.1
+    // @Units: Centidegrees
+    // @User: Advanced
+    AP_GROUPINFO("DRTMIN", 10, AP_SteerController, _deratefactor,        4500),
```
Comment (muramura) [inline libraries/APM_Control/AP_SteerController.cpp]: DRTFCT and DRTMIN, variable name equal ok?

Comment (EShamaev) [inline libraries/APM_Control/AP_SteerController.cpp]: That is typo during update...

**Comment by OXINARF** [comment]
@EShamaev Please read ardupilot.org/dev/docs/submitting-patches-back-to-master.html  You are not following our rules, namely a clean history (no merge commits and squashing fixes) and proper commit messages.

**Comment by EShamaev** [comment]
Should I delete this branch and make a new one (without merging the official ArduPilot/ardupilot)?

**Comment by OXINARF** [comment]
No need to delete the branch, just force push to it. You shouldn't merge, you should rebase your work on master.

**Comment by EShamaev** [comment]
rebased, force pushed

**Comment by OXINARF** [comment]
You haven't rebased on master and deleted the merge commits. Search for interative rebase to learn how to use it. If you need help come to Gitter.

**Comment by EShamaev** [comment]
Like this?

**Comment by OXINARF** [comment]
Yes, that's better. You still haven't the proper commit message, but that's fine for now, just remember it next time.

**Comment by EShamaev** [comment]
Alright, understood about github, fill follow that rules.
But still can't understand what's wrong with message? I corrected and added Subsystem and brief and long description...

**Comment by OXINARF** [comment]
Oh I see what you did, you changed the GitHub pull request title. I'm referring to the commits (in this case you only have one). Your commit message is just:

`Added derating of steering wheel`

but should be:

`APM_Control: added derating of steering wheel` or `AP_SteerController: added derating of steering wheel`

**Comment by EShamaev** [comment]
Understood! Ok, wil follow!

**Comment by EShamaev** [comment]
I based this on my experience.
Also next major update I will show is the taxi waypoints and as it is already done and now under tests, this problem showed there.
We need to know the speed after which we need to derate. Also the amount of derating. And some must be left so we have at least minor control on high speed.
I looked over all parameters, but not much is about steering.

File: libraries/APM_Control/AP_SteerController.cpp [hunk 3]
```diff
@@ -156,11 +184,21 @@ int32_t AP_SteerController::get_steering_out_rate(float desired_rate)
...
-	_last_out = _pid_info.D + _pid_info.FF + _pid_info.P + _pid_info.I;
+    // Calculate the demanded control surface deflection
+    _last_out = _pid_info.D + _pid_info.FF + _pid_info.P + _pid_info.I;
 	
-	// Convert to centi-degrees and constrain
-	return constrain_float(_last_out * 100, -4500, 4500);
+    float derate_constraint = 4500;
+
+    // Calculate required constrain based on speed
+    if (speed > _deratespeed) {
```
Comment (khancyr) [inline libraries/APM_Control/AP_SteerController.cpp]: at _deratespeed = 0 , derating is always applied , this is in contradiction with "If 0 then no derating is used."

Comment (EShamaev) [inline libraries/APM_Control/AP_SteerController.cpp]: Noted, thank you.

**Comment by khancyr** [comment]
@OXINARF we should renable autotest for Rover as this PR will impact Rover too !

**Comment by EShamaev** [comment]
Default values should prevent this.

**Comment by khancyr** [comment]
@EShamaev yep now it is good!

**Comment by OXINARF** [comment]
> @OXINARF we should renable autotest for Rover as this PR will impact Rover too !

Autotest in CI was never enabled for Rover. Feel free to open a PR to add it.

File: libraries/APM_Control/AP_SteerController.cpp [hunk 4]
```diff
@@ -83,6 +83,34 @@ const AP_Param::GroupInfo AP_SteerController::var_info[] = {
...
+    // @Description: Degrees of steering wheel to derate at each additional m/s of speed above "Derating speed". Should be set so that at higher speeds the plane does not roll to the wing in turns.
+    // @Range: 0.0 50.0
+    // @Increment: 0.1
+    // @Units: degree/(m/s)
+    // @User: Advanced
+    AP_GROUPINFO("DRTFCT", 9, AP_SteerController, _deratefactor,        10),
+
+    // @Param: DRTMIN
+    // @DisplayName: Minimum angle of wheel
+    // @Description: The angle that limits smalles angle of steering wheel at maximum speed. Even if it should derate below, it will stop derating at this angle.
```
Comment (gmorph) [inline libraries/APM_Control/AP_SteerController.cpp]: smalles is spelt wrong I believe

**Comment by gmorph** [comment]
These parameters will be visible to Rover users as well.  It would be good to have the descriptions for the parameters suitable for Rover's as well as Plane's - I'm happy to help with the wording.  My understanding of the code is that once the vehicle reaches a certain speed - DRTSPD - turn angle is reduced - DRTFCT - is that correct?  I don't quite follow what DTRMIN is for - could you clarify please?
Thanks, Grant.

**Comment by EShamaev** [comment]
DRTSPD - speed after which angle gets reduced
DRTFCT - amount in degrees to reduce for each additional m/s
DTRMIN - minimum angle that is left

image

**Comment by EShamaev** [comment]
Rebased and pushed to check on rover.

**Comment by gmorph** [comment]
I see - its the minimum angle the steering can be reduced to.  Love that picture - it should go onto the wiki as it makes it very clear what is going on.
LGTM
Thanks, Grant.

**Pull request merged**